packages/date-time-picker/src/vaadin-date-time-picker.js

```diff
@@ -449,87 +469,98 @@ class DateTimePicker extends FieldMixin(SlotMixin(DisabledMixin(ThemableMixin(El
   }
 
   /** @private */
-  __datePickerChanged() {
-    const datePicker = this._getDirectSlotChild('date-picker');
-    if (this.__datePicker === datePicker || !datePicker) {
+  __onDomChange(addedNodes) {
```

Comment:
Using a "slotchange" event should be sufficient for this (FlattenedNodesObserver also uses it under the hood). Tried an alternative fix in https://github.com/vaadin/web-components/commit/ab4308c5ab06bcc221d66a477d163e3975fa3797 At least all the new tests are passing with the approach. Do you see some issues with not switching to FlattenedNodesObserver (which is a Polymer helper)?

Reply:
Looks good I think. I guess the root problem we want to solve is to not just take the first slotted element (as in `_getDirectSlotChild`). I think we would still need to keep the `this.__datePicker === datePicker` check, otherwise we might add event listeners multiple times.

@web-padawan What do you think?

Reply:
One thing that bothers me a bit is that there are a lot of assumptions in the code about the content of the slots. Like which type of element we expect there, and how many of them (0-2). I wonder if we could add some sort of assertions. Also for the Flow use-case, which is imperative, an imperative API like `setDatePicker` might actually make more sense, and is simpler / safer.

Reply:
All in all, the change looks good to me as it is. The two approaches aren't that different at the end of the day.

Reply:
> an imperative API like setDatePicker might actually make more sense, and is simpler / safer

I think this should still work as expected:

```js
const dtp = document.createElement('vaadin-date-time-picker');
document.body.appendChild(dtp);

const dp = document.createElement('vaadin-date-picker');
dp.slot = 'date-picker';
dtp.appendChild(dp);
```